doc/_extensions/zephyr/doxyrunner.py

```diff
@@ -0,0 +1,332 @@
+        else:
+            for pattern in file_patterns:
+                for p_file in path.glob("**/" + pattern):
+                    files.append((p_file, p_file.stat().st_mtime_ns))
+
+    # check if any file has changed
+    dirty = True
+    files_cache_file = cache_dir / "doxygen.cache"
+    if files_cache_file.exists():
+        with open(files_cache_file, "rb") as f:
+            files_cache = pickle.load(f)
+        dirty = files != files_cache
+
+    if not dirty:
+        return False
+
+    # store current state
+    with open(files_cache_file, "wb") as f:
+        pickle.dump(files, f)
+
+    return True
+
+
+def run_doxygen(
+    doxygen: str, doxyfile: str, outdir: Path, silent: bool = False
+) -> None:
+    """Run Doxygen build.
+
+    Args:
+        doxygen: Path to Doxygen binary.
+        doxyfile: Doxyfile content.
+        outdir: Output directory.
+        silent: If Doxygen output should be logged or not.
+    """
+
+    with tempfile.NamedTemporaryFile("w") as f:
+        f.write(doxyfile)
+        f.flush()
+
+        p = Popen([doxygen, f.name], stdout=PIPE, stderr=STDOUT)
+        while True:
+            line = p.stdout.readline()  # type: ignore
+            if p.poll() is not None:
+                break
+            if line and not silent:
+                logger.error(line.decode("utf-8").rstrip())
+
+        if p.returncode:
+            raise IOError(f"Doxygen process returned non-zero ({p.returncode})")
+
+
+def sync_doxygen(doxyfile: str, prev_outdir: Path, latest_outdir: Path) -> None:
+    """Synchronize Doxygen output with a previous build.
+
+    This function makes sure that only new, deleted or changed files are
+    actually modified in the previous Doxygen XML output. Latest HTML content
+    is just moved.
+
+    Args:
+        doxyfile: Contents of the Doxyfile.
+        prev_outdir: Previous Doxygen output directory.
+        latest_outdir: Latest Doxygen build output directory.
+    """
+
+    html_output = get_doxygen_option(doxyfile, "HTML_OUTPUT")
+    if not html_output:
+        raise ValueError("No HTML_OUTPUT set in Doxyfile")
+
+    prev_htmldir = prev_outdir / html_output[0]
+    if prev_htmldir.exists():
+        shutil.rmtree(prev_htmldir)
+
+    latest_htmldir = latest_outdir / html_output[0]
+    latest_htmldir.rename(prev_htmldir)
+
+    xml_output = get_doxygen_option(doxyfile, "XML_OUTPUT")
+    if not xml_output:
+        raise ValueError("No XML_OUTPUT set in Doxyfile")
+
+    prev_xmldir = prev_outdir / xml_output[0]
+    latest_xmldir = latest_outdir / xml_output[0]
+    if prev_xmldir.exists():
+        dcmp = filecmp.dircmp(latest_xmldir, prev_xmldir)
+
+        for file in dcmp.right_only:
+            (Path(dcmp.right) / file).unlink()
+
+        for file in dcmp.left_only + dcmp.diff_files:
+            shutil.copy(Path(dcmp.left) / file, Path(dcmp.right) / file)
+
+        shutil.rmtree(latest_xmldir)
+    else:
+        latest_xmldir.rename(prev_xmldir)
```

Comment:
I think it's better to not anticipate future needs and stay specific with Doxygen/Sphinx needs for now. A full-featured rsync function may imply handling edge cases not needed/covered here. I agree though that we should have a common utility library if in the future other plugins need similar functionality.

Reply:
I wrote "generic", not "full-featured". I'm only suggesting a small code style change, not a new "common utility library". It's much faster to read a one-line `copy_something()` function call and abstraction than infer the same despite `prev_xmldir` and other symbol noise unrelated to some pure copying logic. It may require a bit of effort https://martinfowler.com/bliki/TwoHardThings.html

No big deal.

Reply:
I think the function signature and its docstrings make its purpose quite clear:

```python
def sync_doxygen(doxyfile: str, outdir: Path, latest_outdir: Path) -> None:
    """Synchronize Doxygen output with a previous build.
    This function makes sure that only new, deleted or changed files are
    actually modified in the Doxygen XML output. Latest HTML content is just
    moved.
    Args:
        doxyfile: Contents of the Doxyfile.
        outdir: Doxygen output directory.
        latest_outdir: Latest Doxygen build output directory.
    """
```

I can add some more in-line comments if it helps

Reply:
I think we keep talking across each other so I will be more specific with some code.

First I find the name `latest_outdir` confusing because it makes me think about source control and/or "latest and greatest" and this is neither. In fact the argument name where the function is used is "tmpdir" which sounds pretty different and much more like it. `tmp_builddir` would be better or you could also go with a simple  `sync_doxygen(doxyfile: str, src_dir: Path, dest_dir: Path)` just like `cp`, `rsync`, `memcpy`, `shutil.copy` etc. Swapping the argument order would also be consistency with all these and with your own `filecmp.dircmp` invocation.

Now back to my first suggestion:

```python

 ...

    if dst_xmldir.exists():
        smart_rcopy(src_xmldir, dst_xmldir)
        shutil.rmtree(src_xmldir)
    else:
        src_xmldir.rename(dst_xmldir)

# (Nothing about XML or doxygen in this simple and generic function,
# it's contained above.)
def smart_rcopy(src, dst):
   """Recursive copy that does not copy identical files not to update their
    timestamps"""

        dcmp = filecmp.dircmp(src, dst)

        for file in dcmp.right_only:
            (Path(dcmp.right) / file).unlink()

        for file in dcmp.left_only + dcmp.diff_files:
            shutil.copy(Path(dcmp.left) / file, Path(dcmp.right) / file)
```

Reply:
I have modified the function signature so that it has the src/dest order, but I don't see the real benefit of adding `smart_rcopy`, so I haven't changed that part.

Reply:
> but I don't see the real benefit of adding smart_rcopy

Like so many other things in programming, it's just a small abstraction to help (a little bit) with readability. So the "smart copy + delete XML" part of the logic is not obscured by distracting details about how the smart copy is implemented, and the smart copy implementation that does not care about XML does not have distracting XML symbol names. When I reviewed the code, both these distractions slowed me down enough to care and provide this suggestion. Giving the "smart copy" abstraction an (even better?) name also boosts readability.

"“Programs must be written for people to read, and only incidentally for machines to execute.”

It's a small thing, never mind.